### ISSUE TITLE
File watcher has problems

### ISSUE THREAD

**Issue by contra** [gulpjs/gulp#651]
I'm merging all file watching related issues into one issue, since there is one solution. The next version of gaze is under development and we are waiting for it to come out before we can fix the following issues:

- high cpu usage
- doesnt catch new files
- globs get matched even though it was a file change that shouldn't match it
- new files arent watched
- stops on error (unrelated to gaze, but fixed already in gulp 4)

Please do not open any new issues regarding these problems. The current rate is 3-5 a day, they will be closed for duplication.

@shama is working on fixing these issues in gaze, if you have any new file watching issues you didn't see in the list please open them at https://github.com/shama/gaze

**Comment by darsain** [comment]
Perhaps a lot of current issues would be fixed by updating to newest gaze? Because you are quite behind. Current gulp (3 & 4) install resolves to 0.5.1, but gaze is on 0.6.4 already.

Specifically the *gulp.watch throws error on removing folder* which is what I'm running into right now seems to be already fixed in https://github.com/shama/gaze/issues/44.

**Comment by shama** [comment]
@darsain https://github.com/shama/gaze/issues/137

**Comment by contra** [comment]
@darsain The new gaze is more broken than the old one

**Comment by TrySound** [comment]
Maybe replace `gaze` with [`chokidar`](https://github.com/paulmillr/chokidar)?
I think it's more powerful and flexible.

**Comment by contra** [comment]
@TrySound They need to finish the work on glob support before it becomes feasible, but it is something we will look into when chokidar is ready

**Comment by TrySound** [comment]
Good:)

**Comment by shama** [comment]
![homer-simpson-bush-gif](https://cloud.githubusercontent.com/assets/99604/5841747/1ce1adfa-a154-11e4-8c52-4d761c2f946f.gif)

**Comment by TrySound** [comment]
@contra [1.0.0-rc2](https://github.com/paulmillr/chokidar) was released

**Comment by es128** [comment]
lol @shama, great gif, haven't seen that one before

I did offer to join forces when I was first asked to take the reins on chokidar :smile:: https://github.com/paulmillr/chokidar/issues/138#issuecomment-58018439

but c'mon, you know you still deserve credit for the great work done on gaze

**Comment by es128** [comment]
https://github.com/wearefractal/glob-watcher/pull/13

**Comment by shama** [comment]
@es128 No worries. More options is always better than being stuck with one. If chokidar works better than gaze for you then you should use it. Many libs I use have chokidar as a dep, so I'm grateful for the work you've put into it. :)

Anyways, if anyone needs me, I'll be in my bush.

**Comment by Stickiler** [comment]
Has there been any progress on the high cpu usage issue? I'm attempting to use gulp for a significant project at the moment, but the cpu usage is killing it for me.

**Comment by floatdrop** [comment]
@Stickiler you can use [gulp-watch](https://github.com/floatdrop/gulp-watch) while this issue is in progress and then port back to bundled watch.

**Comment by phated** [comment]
@shama is https://github.com/shama/navelgazer ready for consumption?  I could see us using that without of all of the bells and whistles of gaze.  Thoughts? @contra @heikki @sindresorhus

**Comment by shama** [comment]
@phated Not really. It's not well vetted like chokidar or gaze. It has some duplicate event issues and uses kqueue (which means you'll hit EMFILE a lot sooner than you would with chokidar and gaze@0.5.1). Plus a lot of the issues, at least for me with gaze, are related to watching file patterns more than the actual watch events. So that part would still be difficult if you still wanted file patterns as an interface.

**Comment by phated** [comment]
@shama thanks for the information, really appreciate the insights into this

**Comment by mattfelten** [comment]
Any update on this? Just ran into the error when deleting a folder yesterday.

**Comment by fabpopa** [comment]
Just as a note, while this issue is open, some alternatives are @floatdrop's [gulp-watch](https://github.com/floatdrop/gulp-watch) or [Browsersync.watch()](http://www.browsersync.io/docs/api/#api-watch), both of which are wrappers for [chokidar](https://github.com/paulmillr/chokidar). I'm using the Browsersync method and add/change/delete/rename/delete-directory events work very well with paths as well as globs.

**Comment by igoradamenko** [comment]
@fabpopa So chokidar has some bugs too paulmillr/chokidar#292

**Comment by PyroSA** [comment]
It'll be summer soon in the summer Hemisphere, and this is causing my CPU to act as a heater.
(Windows Environment).

It occasionally does a back-flip and lands on it's head when changing branches as well.

**Comment by contra** [comment]
Status update: going to start working on switching to chokidar now that it has stabilized

**Comment by phated** [comment]
Now that chokidar has been merged into the 4.0 branch. Can we close this?

**Comment by PyroSA** [comment]
Got me all excited, until I realised 4.0 isn't available for our project yet.

**Comment by AnthumChris** [comment]
@mattfelten: I'm catching the errors caused by file renaming and silently ignoring them

```
gulp.task('watch', ['build'], function(cb) {
  gulp.watch('src/**/*.js', { interval: 200 }, ['build'])
    .on('error', function(error) {
      // silently catch 'ENOENT' error typically caused by renaming watched folders
      if (error.code === 'ENOENT') {
        return;
      }
    })
})
```

```
$ node --version
v4.2.2

$ gulp --version
[11:54:01] CLI version 1.0.0
[11:54:01] Local version 3.9.0
```

**Comment by leonardothibes** [comment]
Hello guys.

I found a cause of this problem: many files opened in same time are blocked by S.O.

In Ubuntu/Linux, for example, you can increase this limit editing a file _/etc/security/limits.conf_(you must reboot your Linux after edit this file to apply the changes) and putting this content below in the end of file:

```
soft nofile 4096
hard nofile 4096
```

4096 it's a limit of files with the S.O can open in the same time, you can choice other number for you needs.

This is a S.O. problem, not a gulp problem.

Regards.

**Comment by AlexLeung** [comment]
When I delete a folder that is matched by the glob I send to gulp.watch, 
I still get

```
events.js:154
      throw er; // Unhandled 'error' event
      ^

Error: watch null EPERM
    at exports._errnoException (util.js:890:11)
    at FSEvent.FSWatcher._handle.onchange (fs.js:1296:21)

```

**Comment by spiritix** [comment]
So when will the next version of gaze be released? Anyone knows anything?
Can't really use gulp watch like this, very annoying.

**Comment by phated** [comment]
@spiritix their 1.0 was released but it was a reversion to the code we use from the 0.5 branch so it will never be fixed in gulp 3.x.  Upgrade to gulp 4, switch to chokidar or something else.

I'm locking this thread.
